### PR TITLE
feat(document): document deserialization & serialization

### PR DESCRIPTION
**Pull request by chillleader**
## Description

This PR will add the first implementation of serialization and deserialization of document references.

Highlights:

- Jackson deserializer to handle the simple case when connector wants to get a Document object
- Jackson deserializer to handle more complex cases when target type is not a Document but the document can be converted to the target type
- Operation execution
  - Lazy and eager models are available
  - Lazy is applied by default when the connector doesn't specify its target type (e.g. just Object). The operation will be evaluated on serialization instead in this case.
- Jackson serializer that can execute operations
- Refined document class structure
- DocumentStore interface and in-memory impl

## Related issues

related https://github.com/camunda/team-connectors/issues/873

### CODE REVIEW

**Review by sbuettner — APPROVED** [review]

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/model/HttpCommonRequest.java [hunk 1]
```diff
@@ -79,7 +79,6 @@ public class HttpCommonRequest {
       description = "Map of HTTP headers to add to the request")
   private Map<String, String> headers;
 
-  @FEEL
```
Comment (mathias-vandaele) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/model/HttpCommonRequest.java:82]: Why removing it `@FEEL` ?

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/document/DocumentFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import io.camunda.connector.api.document.DocumentReference.ExternalDocumentReference;
+import io.camunda.connector.api.document.store.DocumentCreationRequest;
+import io.camunda.connector.api.document.store.DocumentStore;
+
+public class DocumentFactoryImpl implements DocumentFactory {
+
+  private final DocumentStore documentStore;
+
+  public DocumentFactoryImpl(DocumentStore documentStore) {
+    this.documentStore = documentStore;
```
Comment (mathias-vandaele) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/document/DocumentFactoryImpl.java:32]: Will this class carry all of the future possible store ? In this case, how do you make the link between the document creation request and the corresponding store?

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/JobHandlerContext.java [hunk 3]
```diff
@@ -47,17 +52,34 @@ public class JobHandlerContext extends AbstractConnectorContext
...
+  public JobHandlerContext(
+      final ActivatedJob job,
+      final SecretProvider secretProvider,
+      final ValidationProvider validationProvider,
+      final ObjectMapper objectMapper) {
+    this(
+        job,
+        secretProvider,
+        validationProvider,
+        new DocumentFactoryImpl(new InMemoryDocumentStore()),
```
Comment (mathias-vandaele) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/JobHandlerContext.java:79]: How will it work when new store comes?

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/model/HttpCommonRequest.java (hunk 1 above)
Comment (chillleader) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/model/HttpCommonRequest.java:82]: Good catch, this is accidental 😄

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/document/DocumentFactoryImpl.java (hunk 2 above)
Comment (chillleader) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/document/DocumentFactoryImpl.java:32]: In the future we will generally use only one store here for connecting to the C8 document API. The in-memory implementation will be reserved for tests. As of now, I don't see the need to have multiple store implementations active at the same time - different stores are anyway hidden behind the C8 REST API.

One thing where a different approach is needed is external documents (such as something only available to us via an external URL). But this probably will be handled separately in the `DocumentFactory`

**Review by mathias-vandaele — APPROVED** [review]

File: connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentFactory.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+package io.camunda.connector.api.document;
+
+import io.camunda.connector.api.document.store.DocumentCreationRequest;
+
+public interface DocumentFactory {
+
+  /** Given a document reference, create the Document object */
+  Document parse(DocumentReference reference);
```
Comment (sbuettner) [inline connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentFactory.java:24]: Maybe `resolve` or `get` would be less technical.

File: connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentReference.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.api.document;
+
+public interface DocumentReference {
```
Comment (sbuettner) [inline connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentReference.java:19]: Should we seal this interface?

**Review by sbuettner — APPROVED** [review]
Great work! Some minor comments.

File: connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentReference.java (hunk 5 above)
Comment (chillleader) [reply connector-sdk/core/src/main/java/io/camunda/connector/api/document/DocumentReference.java:19]: It's being extended / implemented by classes in a different package (`DocumentReferenceImpl` where operation is added) so I couldn't seal it here

**Review by sbuettner — APPROVED** [review]

File: connector-sdk/jackson-datatype-document/src/main/java/io/camunda/connector/document/annotation/jackson/DocumentReferenceModel.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    visible = true,
+    include = As.EXISTING_PROPERTY)
+@JsonSubTypes({
+  @JsonSubTypes.Type(value = CamundaDocumentReferenceModel.class, name = "camunda"),
+  @JsonSubTypes.Type(value = ExternalDocumentReferenceModel.class, name = "external")
+})
+@JsonIgnoreProperties(ignoreUnknown = true)
+public sealed interface DocumentReferenceModel extends DocumentReference {
+
+  String DISCRIMINATOR_KEY = "$documentType";
```
Comment (sbuettner) [inline connector-sdk/jackson-datatype-document/src/main/java/io/camunda/connector/document/annotation/jackson/DocumentReferenceModel.java:47]:
```suggestion
  String DISCRIMINATOR_KEY = "documentType";
```

Should we remove the $-sign as done here: https://github.com/camunda/camunda/pull/21080/files#diff-a18635d76659a781a6e8d301bd0fed7c77fd6cc5c8cfc57e11c5bbc19c78c9cdR1783 ?

**Pull request merged**